pkg/envoy/xds/server.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"errors"
+	"io"
+	"reflect"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/cilium/cilium/pkg/envoy/api"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// AnyTypeURL is the default type URL to use for ADS resource sets.
+	AnyTypeURL = ""
+)
+
+var (
+	// ErrNoADSTypeURL is the error returned when receiving a request without
+	// a type URL from an ADS stream.
+	ErrNoADSTypeURL = errors.New("type URL is required for ADS")
+
+	// ErrUnknownTypeURL is the error returned when receiving a request with
+	// an unknown type URL.
+	ErrUnknownTypeURL = errors.New("unknown type URL")
+
+	// ErrInvalidVersionInfo is the error returned when receiving a request
+	// with a version info that is not a positive integer.
+	ErrInvalidVersionInfo = errors.New("invalid version info")
+
+	// ErrResourceWatch is the error returned whenever an internal error
+	// occurs while waiting for new versions of resources.
+	ErrResourceWatch = errors.New("resource watch failed")
+)
+
+// Server implements the handling of xDS streams.
+type Server struct {
+	// watchers maps each supported type URL to its corresponding resource
+	// watcher.
+	watchers map[string]*ResourceWatcher
+
+	// lastStreamID is the identifier of the last processed stream.
+	// It is incremented atomically when starting the handling of a new stream.
+	lastStreamID uint64
```

Comment:
Apart from being incremented, is this field used?

Reply:
Yes, it's used as a logging field. Cf. the uses of `logfields.XDSStreamID` in `server.go`. This will be useful for debugging since we'll have many long-lived concurrent streams.

Reply:
oops! Totally my bad. I missed that it was assigned to streamID and that was used!